bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/WriteBatchEntryProcessor.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.bookkeeper.net.BookieId;
+import org.apache.bookkeeper.proto.BookieProtocol.ParsedAddRequest;
+import org.apache.bookkeeper.proto.BookkeeperInternalCallbacks.WriteCallback;
+import org.apache.bookkeeper.util.MathUtils;
+
+/**
+ * Processes batched add entry requests.
+ */
+@Slf4j
+public class WriteBatchEntryProcessor extends PacketProcessorBase<ParsedAddRequest> implements WriteCallback {
+    long startTimeNanos;
+    List<ParsedAddRequest> requests;
+    AtomicInteger requestCount = new AtomicInteger(0);
+
+    @Override
+    protected void reset() {
+        requests = null;
+        requestHandler = null;
+        requestProcessor = null;
+        requestCount.set(0);
+        startTimeNanos = -1L;
+    }
+
+    public static WriteBatchEntryProcessor create(List<ParsedAddRequest> requests, BookieRequestHandler requestHandler,
+                                                  BookieRequestProcessor requestProcessor) {
+        WriteBatchEntryProcessor wbep = RECYCLER.get();
+        wbep.init(requests, requestHandler, requestProcessor);
+        requestProcessor.onAddRequestStart(requestHandler.ctx().channel(), requests.size());
+        return wbep;
+    }
+
+    protected void init(List<ParsedAddRequest> requests, BookieRequestHandler requestHandler,
+                      BookieRequestProcessor requestProcessor) {
+        this.requests = requests;
+        this.requestHandler = requestHandler;
+        this.requestProcessor = requestProcessor;
+        this.enqueueNanos = MathUtils.nowInNano();
+        this.requestCount.set(requests.size());
+    }
+
+    @Override
+    protected void processPacket() {
+
+    }
+
+    @Override
+    public void writeComplete(int rc, long ledgerId, long entryId, BookieId addr, Object ctx) {
+        if (BookieProtocol.EOK == rc) {
+            requestProcessor.getRequestStats().getAddEntryStats()
+                .registerSuccessfulEvent(MathUtils.elapsedNanos(startTimeNanos), TimeUnit.NANOSECONDS);
+        } else {
+            requestProcessor.getRequestStats().getAddEntryStats()
+                .registerFailedEvent(MathUtils.elapsedNanos(startTimeNanos), TimeUnit.NANOSECONDS);
+        }
+
+        requestHandler.prepareSendResponseV2(rc, BookieProtocol.CURRENT_PROTOCOL_VERSION, ADDENTRY, ledgerId, entryId);
+        requestProcessor.onAddRequestFinish();
+
+        if (requestCount.decrementAndGet() == 0) {
+            recycle();
+        }
+    }
+
+    @Override
+    public void run() {
+        if (requestProcessor.getBookie().isReadOnly()) {
+                log.warn("BookieServer is running in readOnly mode, so rejecting the request from the client!");
+                for (ParsedAddRequest r : requests) {
+                    writeComplete(BookieProtocol.EREADONLY, r.getLedgerId(), r.getEntryId(), null,
+                        requestHandler.ctx());
+                    r.release();
+                    r.recycle();
+                }
+                return;
+        }
+
```

Comment:
When batch add, do we need to update the `writeThreadQueuedLatency` metric? like: https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PacketProcessorBase.java#L176-L184

Reply:
Good point, added it.